ui/app/pages/permissions-connect/permissions-connect.component.js

```diff
@@ -99,62 +130,39 @@ export default class PermissionConnect extends Component {
     }, () => this.props.history.push(this.props.confirmPermissionPath))
   }
 
-  redirectFlow (accepted) {
-    const { history, location, confirmPermissionPath } = this.props
+  redirectFlow (approved) {
+    const { history } = this.props
 
     this.setState({
       redirecting: true,
-      permissionAccepted: accepted,
+      permissionsApproved: approved,
     })
     this.removeBeforeUnload()
 
+    const timeout = approved ? APPROVE_TIMEOUT : REJECT_TIMEOUT
+
     if (getEnvironmentType() === ENVIRONMENT_TYPE_NOTIFICATION) {
       setTimeout(async () => {
         global.platform.closeCurrentWindow()
-      }, 1500)
-    } else if (location.pathname === confirmPermissionPath) {
+      }, timeout)
+    } else {
       setTimeout(async () => {
```

Comment:
_Note:_ `redirectFlow` has been renamed to `redirect`.

So, I just tested this again, and calling `history.push(DEFAULT_ROUTE)` (directly or via `redirect`) _caused_ the jankyness, in particular if on the second part of the connect form.

I believe this is because if it's called directly, there's still a permission request pending, and the component has to update (with the permissions request now gone), in order to be routed back to `DEFAULT_ROUTE`.

I solved this by only calling `redirectFlow` inside `componentDidUpdate`. What happens when calling `cancelPermissionsRequest` now is:

1. We dispatch the `rejectPermissionsRequest` action.
2. The permissions request is removed.
3. All `permissions-connect` receive new props, without any permissions request. This causes `redirectFlow` to be called in `componentDidUpdate`.
4. The components redirect appropriately.

They now feel responsive enough to me.

Reply:
It sounds like you've just _delayed_ calling `history.push(DEFAULT_ROUTE)` by doing this. I'm not sure I understand how waiting for the next render before calling it can be faster.

Reply:
<details>
<summary>Redirect in "componentDidUpdate"</summary>
<img src="https://user-images.githubusercontent.com/25517051/84402706-a728f480-abb9-11ea-9f8e-b0f9bd2d6ddd.gif" />
</details>

<details>
<summary>Redirect in "cancelPermissionsRequest"</summary>
<img src="https://user-images.githubusercontent.com/25517051/84402694-a2644080-abb9-11ea-8a2a-f16739556386.gif" />
</details>

Reply:
Hmm. I don't understand that second case - it's showing a flash of the list of accounts to connect to, still showing the `origin` of the permission request that should no longer exist :thinking:

Reply:
Yeah, I'm not totally sure why it's doing that, but that's what happens when `this.redirect(false)` is called inside `cancelPermissionsRequest` after `await rejectPermissionsRequest()`.

That's what I was getting at with this:
> I just tested this again, and calling `history.push(DEFAULT_ROUTE)` (directly or via `redirect`) _caused_ the jankyness, in particular if on the second part of the connect form.

Reply:
And so while you're exactly right that we'd expect things to appear less responsive with the changes I made, we in fact see the opposite.